File: src/dataStructures/graphs/wordLadder.rb
require 'pry'
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}

# Strategy: build a directed graph taking the begin_word
# as the root node; use a comparator method to figure out the 
# next vertex(ices) in the graph
def ladder_length(begin_word, end_word, word_list)
    support = Support.new
    src = begin_word
    graph = {}
    graph[src] = []
    target = {}
    src.each_char {|letter| support.src_hash[letter] = :a }
    end_word.each_char {|letter| target[letter] = :a }

    while !word_list.empty?
        word = word_list.shift
        if !support.word_hash[word]
            support.word_hash[word] = {}
            if support.compare(src, word) 
                graph[src] << word 
            else
                word_list << word
            end
        else
            graph[src].each do |vertex|
                src = vertex
                graph[src] = [] if !graph[src]
                support.src_hash = support.word_hash[vertex]
                if support.compare(src, word) 
                    graph[src] << word 
                else
                    word_list << word
                end
            end
        end
    end 
    p graph
end

class Support
    attr_accessor :word_hash, :src_hash

    def initialize()
        @word_hash = {}
        @src_hash = {}
    end

    def compare(a, b)
        miss = 0
        b.each_char do |letter| 
            word_hash[b][letter] = :a
            miss += 1 if !src_hash[letter]
        end
        miss < 2 ? true : false
    end

end





def ladder_length(begin_word, end_word, word_list)
    graph = {}
    hash = {}
    seen = {}
    src = begin_word
    (word_list + [src]).each do |word|
        hash[word] = {}
        graph[word] = []
        word.each_char do |char|
            hash[word][char] = :w
        end
    end
    queue = []
    levels = 1
    return 0 if !hash[end_word]
    while !word_list.empty?
        dest = word_list.shift
        if seen[dest]
            seen = {}
            src = queue.shift
        end
        seen[dest] = true
        if match(src, dest, hash)
            graph[src] << dest
            # Do not push the end_word to queue as it is the last 
            # vertex in graph and does not need to be processed
            queue << dest if dest != end_word   
            # Recycle the end_word until using all of the vertices
            if dest == end_word && !queue.empty?
                word_list << dest
            end
        else
            word_list << dest
        end
    end
    p queue, graph
end


def ladder_length(begin_word, end_word, word_list)
    solution = Solution.new
    graph = {}
    src = begin_word
    (word_list + [src]).each do |word|
        solution.hash[word] = {} 
        solution.hash2[word] = {} 
        graph[word] = []
        word.each_char do |char|
            solution.hash2[word][char] ? solution.hash2[word][char] += 1 : solution.hash2[word][char] = 1
        end
        temp = solution.hash2[word]
        solution.hash[word] = Hash.new.merge(temp)
    end
    solution.hash2.delete src
    return 0 if !solution.hash[end_word]
    queue = [src]
    levels = 1
    done = false
    while !queue.empty?
        break if done
        length = queue.length
        levels += 1
        length.times {|i|
            break if done                
            src = queue.shift
            solution.hash2.keys.each do |key|
                dest = key
                if solution.match(src, dest)
                    graph[src] << dest if dest != begin_word
                    if dest != end_word   
                        queue << dest if dest != begin_word
                        solution.hash2.delete(dest)                        
                    else
                        done = true
                        break
                    end    
                end
            end
            solution.hash2.delete src
        }
    end
    p done ? levels : 0
end

class Solution
    attr_accessor :hash, :hash2
    def initialize
        @hash = {}
        @hash2 = {}
    end
    
    def match(src, dest)
        miss = 0
        src.chars.each_with_index do |char, index|
            if char != dest[index] || hash[dest][char] == 0 
                miss += 1 
            else 
                hash[dest][char] -= 1
            end
        end 
        hash[dest] = Hash.new.merge(hash2[dest])
        miss < 2 ? true : false
    end
end

begin_word = "hit"
end_word = "cog"
word_list = ["hot","dot","dog","lot","log","cog"]


begin_word = "lost"
end_word = "miss"
word_list= ["most","mist","miss","lost","fist","fish"]


begin_word = "leet"
end_word = "code"
word_list = ["lest","leet","lose","code","lode","robe","lost"]




ladder_length(begin_word, end_word, word_list)

File: src/dataStructures/graphs/wordLadder.js
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
// Strategy: initialize an empty queue and push the begin word, and
// a token string that will signal level order traversals (that is, one level down the tree at a time).
// The givens of the problem could be visualized as a tree,
// which head is the begin word, and a path or paths that may or many not
// contain the endword. The goal is to traverse the tree until finding the endword,
// if at all present. 
// Construct a hash containing the words list as keys and "true" boolean as values;
// Construct an array containing the lower case English  alphabet;
// While queue is not empty, dequeue, then perform permutations on deque word
// changing the letter at index i from "a" to "z" (26 permutations at index i)
// , where 0 <= i <= deque word.length;
// compare permutation to endword or pass it to hash map to see if there is a match;
// if permutations is equal to endword, we are done; if match from map, push match to queue, and delete match from map.
// When processing the token, if queue is not empty, push another token;
// Increase the final counter variable to return by 1 after during toekn processing

// Thsi oslution works, but it is non-optimal
 var ladderLength = function(beginWord, endWord, wordList) {
    let alphabet = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'];
    const token = "TOKEN"
    let queue = [beginWord, token];
    let sequenceCount = 1;
    let map = {};
    for(let word of wordList) {
        map[word] = true;
    }

    if(!map[endWord]) return 0;
    delete map[beginWord];
    let permutation;

    while(queue.length) {
        const current = queue.shift();
        if(current === token) {
            sequenceCount += 1;
            if(queue.length) {
                queue.push(token);
            }
            continue;
        }
        
        const splitted = new Array(current.length).fill(0).map((el, index) => current[index]);

        for(let i = 0; i < splitted.length; i++) {
            permutation = splitted.slice();
            for(let j =0; j < alphabet.length; j++) {
                permutation[i] = alphabet[j];
                const permuString = permutation.join('');
                if(permuString !== current) {
                    if(permuString === endWord) {
                        sequenceCount += 1;
                        console.log(sequenceCount)
                        return sequenceCount;
                    } else if(map[permuString]) {
                        queue.push(permuString);
                        delete map[permuString];
                    }
                }
            }
        }
    }
    console.log(sequenceCount)
    return 0;
};


var ladderLength = function(beginWord, endWord, wordList) {
    let map = {};
    const normalLength = beginWord.length;
    let visited = {};
    wordList.push(beginWord);
    let isInList = false;
    for(let word of wordList) {
        for(let i = 0; i < normalLength; i++) {
            const pattern = word.slice(-normalLength, i) + "*" + word.slice(i+1);
            map[pattern] ? map[pattern].push(word) : map[pattern] = [word];
        }
        if(word === endWord) isInList = true;
    }
    if(!isInList) return 0;
    let levelCount = 1;
    let queue = [beginWord];
    visited[beginWord] = true;

    while(queue.length) {
        const levelNodes = queue.length;
        for(let i=0; i < levelNodes; i++) {
            const current = queue.shift();
            if(current === endWord) {
                return levelCount;
            }
            for(let j =0; j < normalLength; j++) {
                const pattern = current.slice(-normalLength, j) + "*" + current.slice(j+1);
                for(let neighbor of map[pattern]) {
                    if(!visited[neighbor]) {
                        queue.push(neighbor);
                        visited[neighbor] = true;
                    }
                }
            }
        }
        levelCount += 1;
    }
    
    return 0;
}


let begin_word = "hit"
let end_word = "cog"
let word_list = ["hot","dot","dog","lot","log","cog"]

// begin_word = "lost"
// end_word = "miss"
// word_list= ["most","mist","miss","lost","fist","fish"]

// begin_word = "leet"
// end_word = "code"
// word_list = ["lest","leet","lose","code","lode","robe","lost"]

ladderLength(begin_word, end_word, word_list);